src/foam/nanos/ruler/RuleActionMergeDecorator.js

```diff
@@ -0,0 +1,72 @@
+/**
+ * @license
+ * Copyright 2021 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+foam.CLASS({
+  package: 'foam.nanos.ruler',
+  name: 'RuleActionMergeDecorator',
+
+  documentation: `
+    Merge calls occurring within mergeDelay period to the delegate rule action.
+
+    Usage: Add RuleActionMergeDecorator to the 'action' of an async rule to
+    merge repeated calls to the underlying rule action. For example,
+
+      {
+        class: 'foam.nanos.ruler.Rule',
+        ...,
+        async: true,
+        action: {
+          class: 'foam.nanos.ruler.RuleActionMergeDecorator',
+          mergeDelay: 1000,
+          delegate: {
+            class: 'MyRuleAction'
+          }
+        }
+      }
+  `,
+
+  implements: [ 'foam.nanos.ruler.RuleAction' ],
+
+  javaImports: [
+    'foam.core.ContextAgent',
+    'foam.core.DirectAgency',
+    'foam.core.X'
+  ],
+
+  properties: [
+    {
+      class: 'FObjectProperty',
+      of: 'foam.nanos.ruler.RuleAction',
+      name: 'delegate'
+    },
+    {
+      class: 'Int',
+      name:  'mergeDelay',
+      units: 'ms'
+    }
+  ],
+
+  methods: [
+    {
+      name: 'applyAction',
+      javaCode: `
+        if ( ! rule.getAsync() || getMergeDelay() <= 0 ) {
+          getDelegate().applyAction(x, obj, oldObj, ruler, rule, agency);
+          return;
+        }
+
+        var key = rule.getId() + ":" + obj.getProperty("id");
+        ((DirectAgency) agency).schedule(x, new ContextAgent() {
```

Comment:
How do you know it is a DirectAgency? I think the schedule method should be on the Agency interface itself. Are there any issues with that?

Reply:
Only async rules can get to here and we always use DirectAgency with async rules https://github.com/kgrgreer/foam3/blob/711d5f1b40167ff301b6517c0314f641a2f66dc3/src/foam/nanos/ruler/Rule.js#L407

And the concern for adding the schedule method to the Agency interface was I'm not sure what would be the appropriate/correct behavior for all the agencies we currently have.

<img width="1126" alt="Screen Shot 2021-11-26 at 10 33 19 PM" src="https://user-images.githubusercontent.com/441658/143666777-dc46f9b9-54d8-4440-9b4d-f433f36683db.png">

Reply:
With recent Interface updates we can add 'schedule' to the Agency interface and a do nothing implementation - which perhaps throws an unsupported operation exception.  
And/or also create an AbstractAgency which our current implementations would extend, providing an implementation for schedule.

Reply:
Fixed in https://github.com/kgrgreer/foam3/pull/1027/commits/396d4f2c185f5049a5efaeb75f443a35a913b841.